File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>

#define MAX 15

using namespace std;

int N;
int board[MAX];
int cnt;

bool CanSet(int row) {
	for (int col = 0; col < row; col++) {
		if (board[row] == board[col] || row - col == abs(board[row] - board[col])) {
			return false;
		}
	}
	return true;
}

void queen(int row) {
	if (row == N) {
		cnt++;
		return;
	}

	for (int col = 0; col < N; col++) {
		board[row] = col;
		if (CanSet(row)) {
			queen(row + 1);
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	cin >> N;

	queen(0);
	printf("%d", cnt);

	return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.js
const fs = require('fs');
const file = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(file).toString().trim().split('\n');

const N = parseInt(input[0]);
let count = 0;
const board = [];

function solution(N) {
  queen(0);
  console.log(count);
}

solution(N);

function queen(row) {
  if (row === N) {
    count++;
    return;
  }

  for (let col = 0; col < N; col++) {
    board[row] = col;
    if (canSet(row)) {
      queen(row + 1);
    }
  }
}

function canSet(row) {
  for (let col = 0; col < row; col++) {
    if (board[row] === board[col] || row - col === Math.abs(board[row] - board[col])) {
      return false;
    }
  }
  return true;
}
